examples/step-63/step-63.cc

```diff
@@ -19,16 +19,1091 @@
+                                         level,
+                                         direction,
+                                         /*dof_wise_renumbering = */ true);
+          }
+        else if (settings.dof_renum == "random")
+          {
+            for (unsigned int level = 0; level < n_levels; ++level)
+              DoFRenumbering::random(dof_handler, level);
+          }
+        else
+          Assert(false, ExcNotImplemented());
+      }
+
+    mg_constrained_dofs.clear();
+    mg_constrained_dofs.initialize(dof_handler);
+
+    std::set<types::boundary_id> dirichlet_boundary_ids = {0, 1};
+    mg_constrained_dofs.make_zero_boundary_constraints(dof_handler,
+                                                       dirichlet_boundary_ids);
+
+    mg_matrices.resize(0, n_levels - 1);
+    mg_matrices.clear_elements();
+    mg_interface_in.resize(0, n_levels - 1);
+    mg_interface_in.clear_elements();
+    mg_interface_out.resize(0, n_levels - 1);
+    mg_interface_out.clear_elements();
+    mg_sparsity_patterns.resize(0, n_levels - 1);
+    mg_interface_sparsity_patterns.resize(0, n_levels - 1);
+
+    for (unsigned int level = 0; level < n_levels; ++level)
+      {
+        {
+          DynamicSparsityPattern dsp(dof_handler.n_dofs(level),
+                                     dof_handler.n_dofs(level));
+          MGTools::make_sparsity_pattern(dof_handler, dsp, level);
+          mg_sparsity_patterns[level].copy_from(dsp);
+          mg_matrices[level].reinit(mg_sparsity_patterns[level]);
+        }
+        {
+          DynamicSparsityPattern dsp(dof_handler.n_dofs(level),
+                                     dof_handler.n_dofs(level));
+          MGTools::make_interface_sparsity_pattern(dof_handler,
+                                                   mg_constrained_dofs,
+                                                   dsp,
+                                                   level);
+          mg_interface_sparsity_patterns[level].copy_from(dsp);
+
+          // We need both interface in and out matrices since our problem is not
+          // symmetric
+          mg_interface_in[level].reinit(mg_interface_sparsity_patterns[level]);
+          mg_interface_out[level].reinit(mg_interface_sparsity_patterns[level]);
+        }
+      }
+  }
+
+
+  template <int dim>
+  template <class IteratorType>
+  void AdvectionProblem<dim>::assemble_cell(const IteratorType &cell,
+                                            ScratchData<dim> &  scratch_data,
+                                            CopyData &          copy_data)
+  {
+    const unsigned int level = cell->level();
+    copy_data.level          = level;
+
+    const unsigned int dofs_per_cell =
+      scratch_data.fe_values.get_fe().dofs_per_cell;
+    copy_data.dofs_per_cell = dofs_per_cell;
+
+    const unsigned int n_q_points =
+      scratch_data.fe_values.get_quadrature().size();
+    copy_data.cell_matrix.reinit(dofs_per_cell, dofs_per_cell);
```

Comment:
move this line up one block -- it logically belongs to the `dofs_per_cell` block before, not the `n_q_point` block you have here.